### PR TITLE
Introduces LLVMCPUTileAndFuse to get rid of CodegenStrategyDriver 

### PR DESCRIPTION
**Pull request by hanhanW**
This is a step towards to retiring CodegenStrategyDriver.

benchmarks: all

### CODE REVIEW

**Comment by hanhanW** [comment]
The static methods in LLVMCPUTileAndFuse.cpp are copied from LinalgTensorCodegenDriver.cpp and https://github.com/openxla/iree/blob/main/llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/Transforms/Transforms.cpp

The implementations will be removed, so I think we don't have to refactor them for now. We're at a transition state.

**Comment by hanhanW** [comment]
This should be an NFC for all CPU pipelines.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUTileAndFuse.cpp [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+    return WalkResult::advance();
+  });
+
+  if (!rootOp || result.wasInterrupted()) {
+    return failure();
+  }
+  return rootOp;
+}
+
+static FailureOr<SmallVector<int64_t>> getTilingSizesFromConfig(
```
Comment (dcaballe) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUTileAndFuse.cpp:71]: Wondering if all these utilities should be moved to a common place...

File: compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUTileAndFuse.cpp [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+  FailureOr<Operation *> rootOp = getRootOp(funcOp);
+  if (failed(rootOp)) return failure();
+
+  return getTileSizes(rootOp.value(), tilingLevel);
+}
+
+/// Starting from `op` walk all operands backwards to find all
+/// potentially fusable operations, i.e. operations that implement
+/// the `TilingInterface`.
+static llvm::SmallDenseSet<Operation *> collectTiledAndFusedOps(
```
Comment (dcaballe) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUTileAndFuse.cpp]: Move `llvm::SmallDenseSet<Operation *>` return to be an argument by reference? This kind of things help a lot with the performance in debug mode.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUTileAndFuse.cpp [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+          result.count(producer))
+        continue;
+      worklist.push_back(producer);
+      result.insert(producer);
+    }
+  }
+  return result;
+}
+
+struct LLVMCPUTileAndFusePass : LLVMCPUTileAndFuseBase<LLVMCPUTileAndFusePass> {
```
Comment (dcaballe) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUTileAndFuse.cpp:105]: High-level doc about what this is doing, esp. mentioning the granularity at which this works now...

**Review by dcaballe — APPROVED** [review]
LG! Some high-level comments.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUTileAndFuse.cpp (hunk 1 above)
Comment (hanhanW) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUTileAndFuse.cpp:71]: Yes, we should move them to a common place, I'd like to do the cleanup later. I think I'll do the refactoring as moving more things out from CodegenStrategy.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUTileAndFuse.cpp [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+  auto funcOp = dyn_cast<func::FuncOp>(op);
+  if (!funcOp) {
+    funcOp = op->getParentOfType<func::FuncOp>();
+  }
+
+  assert(funcOp && "Missing funcOp");
+
+  Operation *rootOp = nullptr;
+  mlir::iree_compiler::IREE::Codegen::LoweringConfigAttr rootLoweringConfig;
+  auto result = funcOp.walk([&](Operation *op) -> WalkResult {
```
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUTileAndFuse.cpp:49]: We probably want to do a post-order walk. IIRC thats what the codegen driver does.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/test/tile_and_fuse.mlir [hunk 5]
```diff
@@ -0,0 +1,34 @@
+// RUN: iree-opt --pass-pipeline="builtin.module(func.func(iree-llvmcpu-tile-and-fuse{tiling-level=0}))" --split-input-file %s | FileCheck %s
+
+func.func @matmul_bias_add(%arg0 : tensor<?x?xf32>, %arg1 : tensor<?x?xf32>, %arg2 : tensor<?xf32>) -> tensor<?x?xf32> {
```
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/test/tile_and_fuse.mlir:3]: Probably need more tests here, I can take care of add more here for the pad fusion case I am looking into.

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]
One minor comment.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUTileAndFuse.cpp (hunk 4 above)
Comment (hanhanW) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUTileAndFuse.cpp:49]: This is the part about how we get lowering config. Yes, CodegenDriver is doing a post-order walk. The logic about it is in the start of `runOnOperation`, i.e, line 205.

Comment (hanhanW) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUTileAndFuse.cpp:49]: wait, please ignore the comment, I was trying something out in line 205.

Comment (hanhanW) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUTileAndFuse.cpp:49]: I think we should use ReverseIterator for accessing the ops, it's done now.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/test/tile_and_fuse.mlir (hunk 5 above)
Comment (hanhanW) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/test/tile_and_fuse.mlir:3]: What kinds of test you're looking for? The pass looks for TilingInterface ops and apply the same transformation. The test contains TilingInterface ops, and the transformations are tested upstream. I think this one is enough for now?

I added a early-skip test case if the tiling sizes are all zeros.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUTileAndFuse.cpp (hunk 4 above)
Comment (MaheshRavishankar) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUTileAndFuse.cpp:49]: I am not sure I follow.

**Review by MaheshRavishankar — APPROVED** [review]

**Pull request merged**